agent/agent.go

```diff
@@ -34,26 +35,32 @@ type Agent struct {
 	cancelF context.CancelFunc
 	closing uint32
 	ctx     context.Context
-	sigs    chan os.Signal
 
 	provider providers.ExecutionProvider
 
 	cacheBucket nats.ObjectStore
 	md          *agentapi.MachineMetadata
 	nc          *nats.Conn
 	started     time.Time
+
+	sandboxed bool
 }
 
-// HaltVM stops the firecracker VM
+// HaltVM stops the firecracker VM (or the agent if it is not sandboxed)
 func HaltVM(err error) {
+	code := 0
 	if err != nil {
-		// On the off chance the agent's log is captured from the vm
-		fmt.Fprintf(os.Stderr, "Terminating Firecracker VM due to fatal error: %s\n", err)
+		fmt.Fprintf(os.Stderr, "Terminating process due to fatal error: %s. Sandboxed: %v\n", err, isSandboxed())
+		code = 1
 	}
 
-	err = syscall.Reboot(syscall.LINUX_REBOOT_CMD_RESTART)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Failed to reboot: %s", err)
+	if isSandboxed() {
+		err = syscall.Reboot(syscall.LINUX_REBOOT_CMD_RESTART)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to halt: %s", err)
+		}
+	} else {
+		os.Exit(code)
```

Comment:
I am not a huge fan of calling `os.Exit` as it prevents graceful cleanup. Still reviewing this PR though so will hold any suggestions on how we might do it differently for later.

Reply:
this os.Exit is actually called inside the graceful cleanup handler

Reply:
FWIW, os.Exit doesn't allow any pending `defer` to execute.  we should probably be using `runtime.Goexit` https://pkg.go.dev/runtime#Goexit